### PR TITLE
[TE] authentication backend fixes

### PR DESCRIPTION
**Pull request by apucher**
The existing authentication logic has issues for several corner cases with expiring cookies, etc. The PR attempts to resolve them. Furthermore, it includes pointers for necessary changes when using database-persistent user-sessions in the future.

* clean separation between credentials and auth cookie

* caching authenticator

* persist user credentials in encrypted cookie for re-auth if necessary

* add config options for cache and cookie TTL

### CODE REVIEW

File: thirdeye/thirdeye-pinot/src/main/java/com/linkedin/thirdeye/auth/AuthCookieSerializer.java [hunk 1]
```diff
@@ -0,0 +1,40 @@
...
+import java.security.Key;
+import javax.crypto.Cipher;
+import org.apache.commons.codec.binary.Base64;
+import org.codehaus.jackson.map.ObjectMapper;
+import org.codehaus.jackson.type.TypeReference;
+
+
+public class AuthCookieSerializer {
+  final Key aesKey;
+  final ObjectMapper mapper;
```
Comment (cyenjung) [inline thirdeye/thirdeye-pinot/src/main/java/com/linkedin/thirdeye/auth/AuthCookieSerializer.java]: ObjectMapper is a heavy object but thread-safe. Make it final static and be shared across instances.

File: thirdeye/thirdeye-pinot/src/main/java/com/linkedin/thirdeye/auth/ThirdEyeAuthenticator.java [hunk 2]
```diff
@@ -0,0 +1,82 @@
...
+        return Optional.of(principal);
+      }
+
+      // TODO add support for authentication via DB token
+
+    } catch (Exception e) {
+      throw new AuthenticationException(e);
+    }
+
+    LOG.info("Authentication failed for {}", credentials.getPrincipal());
```
Comment (cyenjung) [inline thirdeye/thirdeye-pinot/src/main/java/com/linkedin/thirdeye/auth/ThirdEyeAuthenticator.java]: credentials.getPrincipal() could be null.

File: thirdeye/thirdeye-pinot/src/main/java/com/linkedin/thirdeye/auth/ThirdEyeAuthenticator.java [hunk 3]
```diff
@@ -0,0 +1,82 @@
...
+
+        ThirdEyePrincipal principal = new ThirdEyePrincipal();
+        principal.setName(credentials.getPrincipal());
+
+        LOG.info("Successfully authenticated {}", credentials.getPrincipal());
+        setCurrentPrincipal(principal);
+        return Optional.of(principal);
+      }
+
+      // TODO add support for authentication via DB token
```
Comment (cyenjung) [inline thirdeye/thirdeye-pinot/src/main/java/com/linkedin/thirdeye/auth/ThirdEyeAuthenticator.java]: And handle if(credentials.getPrincipal() == null)

File: thirdeye/thirdeye-pinot/src/main/java/com/linkedin/thirdeye/auth/AuthCookieSerializer.java (hunk 1 above)
Comment (apucher) [reply thirdeye/thirdeye-pinot/src/main/java/com/linkedin/thirdeye/auth/AuthCookieSerializer.java]: AuthCookieSerializer is instantiated by the dashboard app, and dependencies are injected.

File: thirdeye/thirdeye-pinot/src/main/java/com/linkedin/thirdeye/auth/ThirdEyeAuthenticator.java (hunk 3 above)
Comment (apucher) [reply thirdeye/thirdeye-pinot/src/main/java/com/linkedin/thirdeye/auth/ThirdEyeAuthenticator.java]: there's an explicit principal == null check above

File: thirdeye/thirdeye-pinot/src/main/java/com/linkedin/thirdeye/auth/ThirdEyeAuthenticator.java (hunk 2 above)
Comment (apucher) [reply thirdeye/thirdeye-pinot/src/main/java/com/linkedin/thirdeye/auth/ThirdEyeAuthenticator.java]: that's expected and tolerated

File: thirdeye/thirdeye-pinot/src/main/java/com/linkedin/thirdeye/auth/AuthCookieSerializer.java (hunk 1 above)
Comment (apucher) [reply thirdeye/thirdeye-pinot/src/main/java/com/linkedin/thirdeye/auth/AuthCookieSerializer.java]: i've changed the constructor to clarify this

**Pull request merged**